vllm/model_executor/models/pixtral.py

```diff
@@ -576,3 +581,479 @@ def __init__(self, args: VisionEncoderArgs, dim: int):
+                                      image_height: int) -> Tuple[int, int]:
+    # Adapted from transformers.models.pixtral.image_processing_pixtral.get_resize_output_image_size # noqa: E501
+    # https://github.com/huggingface/transformers/blob/2bd4d5897dc73e8b172832070a6f9e567a0df017/src/transformers/models/pixtral/image_processing_pixtral.py#L180 # noqa: E501
+    max_width, max_height = hf_config.image_size, hf_config.image_size
+    patch_width, patch_height = hf_config.patch_size, hf_config.patch_size
+
+    ratio = max(image_width / max_width, image_height / max_height)
+
+    if ratio > 1:
+        image_width = int(numpy.ceil(image_width / ratio))
+        image_height = int(numpy.ceil(image_height / ratio))
+
+    num_height_tokens, num_width_tokens = _num_image_tokens(
+        (image_height, image_width), (patch_height, patch_width))
+
+    return num_width_tokens, num_height_tokens
+
+
+def input_processor_for_pixtral_hf(
+    model_config: ModelConfig,
+    hf_config: PixtralVisionConfig,
+    inputs: DecoderOnlyInputs,
+    *,
+    image_token_id: int,
+    image_feature_size_override: Optional[Union[int, List[int]]] = None,
+) -> DecoderOnlyInputs:
+    assert image_feature_size_override is None, (
+        "image_feature_size_override is not supported for Pixtral")
+
+    multi_modal_data = inputs.get("multi_modal_data")
+    if multi_modal_data is None or "image" not in multi_modal_data:
+        return inputs
+
+    tokenizer = cached_get_tokenizer(model_config.tokenizer)
+    processor = cached_get_processor(model_config.model)
+
+    image_data = multi_modal_data["image"]
+    if isinstance(image_data, Image.Image):
+        image_data = [image_data]
+    elif not is_list_of(image_data, Image.Image):
+        raise TypeError(f"Invalid image type: {type(image_data)}")
+
+    replace_strings = []
+    new_prompt = inputs.get("prompt")
+    for image in image_data:
+        w, h = image.size
+
+        num_width_tokens, num_height_tokens = get_pixtral_hf_image_feature_size(
+            hf_config, image_width=w, image_height=h)
+
+        replace_tokens = [[processor.image_token] * num_width_tokens +
+                          [processor.image_break_token]] * num_height_tokens
+        # Flatten list
+        replace_tokens = [
+            item for sublist in replace_tokens for item in sublist
+        ]
+        replace_tokens[-1] = processor.image_end_token
+        replace_str = "".join(replace_tokens)
+        replace_strings.append(replace_str)
+        new_prompt = new_prompt.replace(processor.image_token, "<placeholder>",
+                                        1)
+
+    while "<placeholder>" in new_prompt:
+        replace_str = replace_strings.pop(0)
+        new_prompt = new_prompt.replace("<placeholder>", replace_str, 1)
+
+    new_token_ids = tokenizer(new_prompt)["input_ids"]
+
+    # NOTE: Create a defensive copy of the original inputs
+    return DecoderOnlyInputs(prompt_token_ids=new_token_ids,
```

Comment:
```suggestion
    return DecoderOnlyInputs(prompt_token_ids=new_token_ids,
```

We should use `token_inputs` instead of `DecoderOnlyInputs` now.